clab/config/template.go

```diff
@@ -0,0 +1,203 @@
+package config
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/srl-labs/containerlab/types"
+)
+
+type ConfigSnippet struct {
+	TargetNode *types.Node
+	// the Rendered template
+	Data []byte
+	// some info for tracing/debugging
+	templateName, source string
+	// All the variables used to render the template
+	vars *map[string]string
+}
+
+// internal template cache
+var templates map[string]*template.Template
+
+func LoadTemplate(kind, templatePath string) error {
+	if templates == nil {
+		templates = make(map[string]*template.Template)
+	}
+	if _, ok := templates[kind]; ok {
+		return nil
+	}
+
+	tp := filepath.Join(templatePath, kind, "*.tmpl")
+	log.Debugf("Load templates from: %s", tp)
+
+	ct := template.New(kind).Funcs(funcMap)
+	var err error
+	templates[kind], err = ct.ParseGlob(tp)
+	if err != nil {
+		log.Errorf("Could not load %s %s", tp, err)
+		return err
+	}
+	return nil
+}
+
+func (c *ConfigSnippet) Render() error {
+	t := templates[c.TargetNode.Kind]
+	buf := new(strings.Builder)
+	c.Data = nil
+
+	varsP, err := json.MarshalIndent(c.vars, "", "  ")
+	if err != nil {
+		varsP = []byte(fmt.Sprintf("%s", c.vars))
+	}
+
+	err = t.ExecuteTemplate(buf, c.templateName, c.vars)
+	if err != nil {
+		log.Errorf("could not render template %s: %s vars=%s\n", c.String(), err, varsP)
+		return fmt.Errorf("could not render template %s: %s", c.String(), err)
+	}
+
+	// Strip blank lines
+	res := strings.Trim(buf.String(), "\n")
+	res = strings.ReplaceAll(res, "\n\n\n", "\n\n")
+	c.Data = []byte(res)
+
+	return nil
+}
+
+func RenderNode(node *types.Node) ([]ConfigSnippet, error) {
```

Comment:
I wonder can we hugely improve the way we do rendering by actually merging the variable/labels of a node and the links that has a leg in this node

What i find quite complicated about the current approach is that we demand to have `*-node.tmpl` and `*-link.tmpl` and render two different templates per node.
If we merge the labels of a node and all its links, we will be able to just render the node and has a single template `$name.tmpl` to render.

I think this greatly improves the simplicity. What is your opinion on that?

Reply:
I like this idea, it makes life simpler. (less indirection and rules to remember)

The only drawback I see in this approach is that you cannot incrementally apply `-node` and then individual `-links` (today I'll have 5 transactions if I have 1 node with 4 links)

but in the overall scheme of things (and for simplicity sake, its a good compromise).

Reply:
> The only drawback I see in this approach is that you cannot incrementally apply `-node` and then individual `-links`

I guess if your workflow needs separation between node and links portions, you would still be able to split the templates in two files and just ignore the nodes variable for link template (and vice versa)